src/humanloop/decorators/tool.py

```diff
@@ -0,0 +1,189 @@
+import builtins
+import inspect
+import textwrap
+import typing
+import uuid
+from typing import Callable, Optional, Union
+from functools import wraps
+
+
+from humanloop.otel import get_trace_context, get_tracer, pop_trace_context, push_trace_context
+from humanloop.otel.constants import HL_FILE_OT_KEY, HL_LOG_OT_KEY, HL_TRACE_METADATA_KEY
+from humanloop.otel.helpers import write_to_opentelemetry_span
+
+from .helpers import args_to_inputs
+
+
+def _type_to_schema(type_hint):
+    if type_hint is int:
+        return "number"
+    if type_hint is float:
+        return "number"
+    if type_hint is bool:
+        return "boolean"
+    if type_hint is str:
+        return "string"
+    if type_hint is dict:
+        return "object"
+    raise ValueError(f"Unsupported type hint: {type_hint}")
+
+
+def _handle_dict_annotation(parameter: inspect.Parameter) -> dict[str, object]:
+    try:
+        type_key, type_value = typing.get_args(parameter.annotation)
+    except ValueError:
+        raise ValueError("Dict annotation must have two type hints")
+    if type_key not in (builtins.str, builtins.int, typing.Literal, builtins.float):
+        raise ValueError("Dict keys must be strings or integers", parameter.name, type_key)
+    if type_value not in (
+        builtins.str,
+        builtins.int,
+        typing.Literal,
+        builtins.float,
+        dict,
+    ):
+        raise ValueError("Dict values must be strings or integers", parameter.name, type_value)
+    return {
+        "type": "object",
+        "properties": {
+            "key": {"type": _type_to_schema(type_key)},
+            "value": {"type": _type_to_schema(type_value)},
+        },
+    }
+
+
+def _handle_list_annotation(parameter: inspect.Parameter) -> dict[str, object]:
+    try:
+        list_type = typing.get_args(parameter.annotation)[0]
+    except ValueError:
+        raise ValueError("List annotation must have one type hint")
+    return {
+        "type": "array",
+        "items": {"type": _type_to_schema(list_type)},
+    }
+
+
+def _handle_union_annotation(parameter: inspect.Parameter) -> dict[str, object]:
+    union_types = [sub_type for sub_type in typing.get_args(parameter.annotation) if sub_type != type(None)]
```

Comment:
These annotation handling functions to convert to json_schema feel very brittle - are you trying to create a full parser here?

Reply:
yes